[4.x] Feat: Rich editor table extensions

schmeits opened the pull request:
## Description

Adds the following extensions to the Rich Editor:

- Table
- Remove Table 
- Add Column Before
- Add Column After
- Remove Column
- Add Row Before
- Add Row After
- Delete Row
- Merge Cells
- Split Cells
- Toggle Header Row

## Visual changes

![image](https://github.com/user-attachments/assets/d08c5f69-30c3-4e4b-b92f-0eb3652047c9)

## Functional changes

- [X] Code style has been fixed by running the `composer cs` command.
- [X] Changes have been tested to not break existing functionality.
- [X] Documentation is up-to-date.

schmeits committed:
Initial commit

First version of the table extension

schmeits committed:
Remove redundant resizable option in TableKit config

The 'resizable' option was set at the top level of TableKit.configure, but it is already specified within the 'table' property. This change removes the redundant configuration.

schmeits committed:
Added dark mode support and use primary colors

schmeits committed:
Added the rest of the functions

Adding/deleting rows, columns, merge cells, split cells and toggle header

schmeits committed:
Too many default buttons

schmeits committed:
composer cs

schmeits committed:
Updated the docs

schmeits committed:
Made a first try implementing a bubble menu

schmeits commented:
You can now use Bubblemenu's using the following syntax

```PHP
Forms\Components\RichEditor::make('bio')
                    ->toolbarButtons([
                        ['bold', 'italic', 'underline', 'strike', 'subscript', 'superscript', 'link'],
                        ['h2', 'h3'],
                        ['blockquote', 'codeBlock', 'bulletList', 'orderedList'],
                        ['table' => [
                            'tableRemove',
                            'tableColumnAddBefore',
                            'tableColumnDelete',
                            'tableColumnAddAfter',
                            'tableRowAddBefore',
                            'tableRowDelete',
                            'tableRowAddAfter',
                            'tableHeaderToggle',
                        ]]]
                    )
```

All the array values are converted to Bubblemenu items

![image](https://github.com/user-attachments/assets/10ca852a-ebee-414e-9f1c-fa16e1939f04)

you could even do this

```PHP
Forms\Components\RichEditor::make('bio')
                    ->toolbarButtons([
                        ['bold' => ['italic', 'underline'], 'italic', 'underline', 'strike', 'subscript', 'superscript', 'link'],
```

![image](https://github.com/user-attachments/assets/1368c448-5a2a-478a-9f47-3382db3c65b1)

danharrin committed:
rebuild

schmeits committed:
Update toolbar button return type annotations

Refines the PHPDoc return type annotations for toolbar button methods in InteractsWithToolbarButtons, MarkdownEditor, and RichEditor to more accurately reflect the returned array structure.

danharrin reviewed (commented):
Can you please work out an API for floating toolbars that does not use the main toolbars API? Like `->floatingToolbars(['table' => [.......]])`

A user could then register a floating toolbar with something like `->floatingToolbars(['text' => [.......]])` which would be active whenever the user selects text, instead of needing to tie it to "bold" for example

I consider these separate toolbars from the main toolbar so I think it makes sense that their definition is separate

schmeits committed:
Changed the API for FloatingToolbars

schmeits commented:
New suggestion:

```PHP
              Forms\Components\RichEditor::make('bio')
                    ->toolbarButtons([
                        ['bold', 'italic', 'underline', 'strike', 'subscript', 'superscript', 'link'],
                        ['h2', 'h3'],
                        ['table'],
                        ['blockquote', 'codeBlock', 'bulletList', 'orderedList'],
                    ])
                    ->floatingToolbars([
                        'heading' => [
                            'h1', 'h2', 'h3',
                        ],
                        'paragraph' => [
                            'bold', 'italic', 'underline', 'strike', 'subscript', 'superscript',
                        ],
                        'table' => [
                            'tableRemove',
                            'tableColumnAddBefore', 'tableColumnDelete', 'tableColumnAddAfter',
                            'tableRowAddBefore', 'tableRowDelete', 'tableRowAddAfter',
                            'tableCellMerge', 'tableCellSplit',
                            'tableHeaderToggle',
                        ]
                    ])
```

![image](https://github.com/user-attachments/assets/2ae46710-50cf-4fab-ab0c-c62685833322)

![image](https://github.com/user-attachments/assets/37481294-8d79-48ac-9c4f-bf68ed9031e2)

![image](https://github.com/user-attachments/assets/2e240bc1-8d43-4baf-a9ad-b1503c3275cc)

schmeits commented:
The key is the node to trigger the floating menu

schmeits committed:
Update InteractsWithToolbarButtons.php

schmeits committed:
Only Richeditor has Floating Toolbars

schmeits committed:
Updated docs

schmeits committed:
Merge remote-tracking branch 'origin/4.x' into rich-editor-table-extensions

schmeits committed:
fixed branch conflicts

schmeits committed:
Merge branch '4.x' into rich-editor-table-extensions

schmeits commented:
Maybe we can fix the dropdown menu's like the ones in https://fluxui.dev/components/editor if we can come up with a nice API for submenu items... would be nice for textAlign and Table functions

danharrin commented:
First thing I notice is that when there is a paragraph floating menu, it is open when the page loads even if the editor is out of focus

<img width="1033" alt="Screenshot 2025-06-26 at 13 38 53" src="https://github.com/user-attachments/assets/85f592d9-47c0-4777-94b7-a6a8fa3299d5" />

There also seems to be overlap/z-index issues with the menu in various places

<img width="1033" alt="Screenshot 2025-06-26 at 13 39 53" src="https://github.com/user-attachments/assets/f186661b-0def-44ff-b9c1-91b33edb81f4" />

I think we need to add some tooltips to the floating menu items, thoughts?

schmeits commented:
what kind of control is that in front of the floating menu?
![image](https://github.com/user-attachments/assets/81501ed3-c7cf-4cbf-888c-68882ba2c480)

schmeits committed:
Only show the bubblemenu if the editor is active

danharrin commented:
> what kind of control is that in front of the floating menu?

That is a navigation item in the sidebar

schmeits committed:
z-ordering fix floating menu

danharrin commented:
Can you add tooltips for all toolbar buttons please?

Also, I think it is probably worth moving the floating toolbars to BELOW the cursor instead of above, so they will never overlap the main toolbar

schmeits committed:
toolbar placement and tooltips

- Align floating toolbar bottom
- Added x-tooltip to all the RichEditorTools (removed the title)

schmeits committed:
using merge attributes is better and also respecting theme

danharrin committed:
Merge branch '4.x' into pr/16676

danharrin committed:
recompile

danharrin committed:
Merge branch 'rich-editor-table-extensions' of https://github.com/schmeits/filament into pr/16676

danharrin committed:
recompile again

danharrin committed:
clean up

danharrin committed:
clean up

danharrin committed:
clean up

danharrin committed:
css

danharrin committed:
Update RichContentRenderer.php

danharrin committed:
rebuild screenshots

danharrin commented:
Thanks!

schmeits commented:
Thank you! Learned a lot from your changes just now... next time I'll try to come closer to your coding style ;)
